File: packages/shared/src/ipc/contracts/aiGenerations.ts
import type { ApplicationAnswer, InterviewQuestion } from '../../types/index.js';

export interface AiGenerationRecord {
  id: string;
  createdAt: number;
  candidateName: string;
  jobTitle: string;
  companyName: string;
  resumeLanguage: string;
  jobAdLanguage: string;
  targetLanguage: string;
  mismatch: boolean;
  topRequirements: string[];
  mode: string;
  resumeText: string;
  coverLetterText: string;
  jobAd: string;
  /** The job this generation targets — links the record to an autopilot found job. */
  jobUrl: string;
  /** The board the job came from (e.g. "linkedin"). */
  board: string;
  /** Answered application questions (the questions assistant), if any. */
  applicationAnswers: ApplicationAnswer[];
  /** The company-research brief used for this application, if any. */
  companyBrief: string;
  /** AI-suggested questions the candidate can ASK the interviewer, if any. */
  interviewQuestions: InterviewQuestion[];
  /**
   * The persisted apply-by-email draft (subject line + body). Optional because
   * records serialised before these columns existed — e.g. an older exported
   * backup replayed through a test fixture — carry neither field.
   */
  emailSubject?: string;
  emailBody?: string;
  /**
   * Parent Application FK — set at save time (and backfilled at boot for legacy
   * rows). The Application detail page joins this generation's docs by this id, not
   * by url, because the Application stores the NORMALIZED url and the generation the
   * RAW one (they never match for query-id boards like Indeed). Absent when unlinked.
   */
  applicationId?: string;
  /**
   * Deterministic content-quality report (serialized `ContentReport` JSON from
   * `validate::content::validate_content`) for the generation that most
   * recently wrote `resumeText`/`coverLetterText`. Absent/empty means no report
   * has been computed yet (or the row predates this field). See ADR-007
   * addendum — a manual text edit via {@link AiGenerationUpdateRequest}
   * deliberately never clears this.
   */
  qualityReport?: string;
}

export interface AiGenerationSaveRequest {
  candidateName: string;
  jobTitle: string;
  companyName: string;
  resumeLanguage: string;
  jobAdLanguage: string;
  targetLanguage: string;
  mismatch: boolean;
  topRequirements: string[];
  mode: string;
  resumeText: string;
  coverLetterText: string;
  jobAd: string;
  /** The job this generation targets (marks the autopilot found job "applied"). */
  jobUrl?: string;
  /** The board the job came from. */
  board?: string;
  /** Answered application questions to persist on the (per-job) record. */
  applicationAnswers?: ApplicationAnswer[];
  /** The company-research brief used, persisted for audit. */
  companyBrief?: string;
  /** AI-suggested interview questions to persist on the (per-job) record. */
  interviewQuestions?: InterviewQuestion[];
  /**
   * The apply-by-email draft to persist on the (per-job) record. Merged like
   * `coverLetterText`: a non-blank value overwrites the stored draft, a blank
   * one leaves it untouched — so a résumé/answers save can't wipe the email.
   */
  emailSubject?: string;
  emailBody?: string;
  /**
   * Deterministic content-quality report to persist alongside this save (see
   * {@link AiGenerationRecord.qualityReport}). Renderer's job to compute it
   * (typically right after a resume/cover regeneration) and pass it here; an
   * absent/empty value leaves whatever report is already on the aggregate.
   */
  qualityReport?: string;
}

/**
 * Edit the résumé/cover-letter text of an existing saved generation, selected by
 * `id`. Unlike {@link AiGenerationSaveRequest} (a per-job merge-upsert that keeps
 * existing non-empty text), this is a direct overwrite — so a user editing a
 * saved generation can blank out or fully replace the text. Each text field is
 * optional; an absent field is left unchanged.
 */
export interface AiGenerationUpdateRequest {
  id: string;
  resumeText?: string;
  coverLetterText?: string;
}

/**
 * Result of `save` — the Rust command reports failure IN-BAND (it resolves with
 * `{ error }` instead of rejecting), so a caller's `onError` never fires for a
 * store failure. Modelled as a union of the two disjoint arms the command
 * actually returns, which makes the compiler refuse a bare `result.id` until
 * the failure arm has been narrowed out (`'error' in result`) — the check is
 * otherwise trivially forgotten, which is exactly how it was missed here.
 */
export type AiGenerationSaveResult = { id: string; success: true } | { error: string };

export interface AiGenerationsContract {
  list(): Promise<AiGenerationRecord[]>;
  save(req: AiGenerationSaveRequest): Promise<AiGenerationSaveResult>;
  update(req: AiGenerationUpdateRequest): Promise<void>;
  remove(id: string): Promise<void>;
  removeBulk(ids: string[]): Promise<void>;
}

export const AI_GENERATIONS_CHANNELS = {
  list: 'aiGenerations:list',
  save: 'aiGenerations:save',
  update: 'aiGenerations:update',
  remove: 'aiGenerations:remove',
  removeBulk: 'aiGenerations:removeBulk',
} as const;

File: apps/tauri/src-tauri/src/commands/ai_generations.rs
use serde_json::{json, Value};
use tauri::{AppHandle, Manager};

use crate::ipc_contracts::ai::{AiGenerationSaveRequest, AiGenerationUpdateRequest};

#[tauri::command]
pub async fn ai_generations_list(app: AppHandle) -> Value {
    let store = app.state::<crate::ai_generations::AiGenerationStore>();
    serde_json::to_value(store.list()).unwrap_or(json!([]))
}

#[tauri::command]
pub async fn ai_generations_save(app: AppHandle, req: AiGenerationSaveRequest) -> Value {
    let store = app.state::<crate::ai_generations::AiGenerationStore>();

    let rec = crate::ai_generations::AiGenerationRecord {
        id: crate::ai_generations::make_generation_id(),
        created_at: crate::ai_generations::now_ms(),
        candidate_name: req.candidate_name,
        job_title: req.job_title,
        company_name: req.company_name,
        resume_language: req.resume_language,
        job_ad_language: req.job_ad_language,
        target_language: req.target_language,
        mismatch: req.mismatch,
        top_requirements: req.top_requirements,
        mode: req.mode,
        resume_text: req.resume_text,
        cover_letter_text: req.cover_letter_text,
        job_ad: req.job_ad,
        job_url: req.job_url,
        board: req.board,
        application_answers: req
            .application_answers
            .into_iter()
            .map(|a| crate::ai_generations::ApplicationAnswer {
                id: a.id,
                question: a.question,
                answer: a.answer,
            })
            .collect(),
        company_brief: req.company_brief,
        interview_questions: req
            .interview_questions
            .into_iter()
            .map(|q| crate::ai_generations::InterviewQuestion {
                id: q.id,
                question: q.question,
                why: q.why,
                audience: q.audience,
            })
            .collect(),
        application_id: None,
    };

    // ADR 0001: the Application aggregate is the source of truth for status +
    // "applied". A generation is its child Document. So before persisting the
    // generation we upsert/advance the Application for this job_url (Generate
    // origin → `applied`), then store the generation (which still carries its
    // own job/company/board copy for backward compatibility + offline export).
    let job_url = rec.job_url.clone();
    let board = rec.board.clone();
    let meta = crate::applications::ApplicationMeta {
        company: rec.company_name.clone(),
        title: rec.job_title.clone(),
        candidate: rec.candidate_name.clone(),
        brief: rec.company_brief.clone(),
        job_description: String::new(), // ponytail: JD persistence is scoped to import + the update IPC
        answers: rec.application_answers.clone(),
        job_summary: String::new(),
    };
    if let Some(apps) = app.try_state::<crate::applications::ApplicationStore>() {
        if let Err(e) = apps.upsert_for_origin(
            &job_url,
            &board,
            &meta,
            crate::applications::ApplicationOrigin::Generate,
            None,
        ) {
            // Non-fatal: a failed Application upsert must not lose the generation
            // the user just produced. The generation save below is the user-visible
            // action; the aggregate can be re-derived.
            log::warn!("[ai_generations] application upsert failed (non-fatal): {e}");
        }
    }

    // Per-job aggregate (generation side): merge into that job's generation row so
    // résumé/cover/answers/brief from separate actions land on one document record.
    match store.save_application(rec) {
        Ok(id) => json!({ "id": id, "success": true }),
        Err(e) => json!({ "error": e }),
    }
}

#[tauri::command]
pub async fn ai_generations_update(app: AppHandle, req: AiGenerationUpdateRequest) -> Value {
    let store = app.state::<crate::ai_generations::AiGenerationStore>();
    // Direct overwrite of exactly the provided text fields, selected by id —
    // distinct from the save merge-upsert, so a user edit can replace text the
    // merge would have kept.
    match store.update_texts(&req.id, req.resume_text, req.cover_letter_text) {
        Ok(()) => json!({ "success": true }),
        Err(e) => json!({ "error": e }),
    }
}

#[tauri::command]
pub async fn ai_generations_remove(app: AppHandle, id: String) -> Value {
    let store = app.state::<crate::ai_generations::AiGenerationStore>();
    match store.remove(&id) {
        Ok(()) => json!({ "success": true }),
        Err(e) => json!({ "error": e }),
    }
}

#[tauri::command]
pub async fn ai_generations_remove_bulk(app: AppHandle, ids: Vec<String>) -> Value {
    let store = app.state::<crate::ai_generations::AiGenerationStore>();
    match store.remove_many(&ids) {
        Ok(count) => json!({ "success": true, "count": count }),
        Err(e) => json!({ "error": e }),
    }
}
